### PR TITLE
Allow updating Calendar props to reset the month and year

### PR DESCRIPTION
**Pull request by jonniedarko**
This should deal with the issue #125 . if you want any changes to accept please let me know, I'd be happy to. styling on example page may be somewhere you might consider requiring an update

### CODE REVIEW

File: dist/css/react-calendar.css [hunk 1]
```diff
@@ -215,7 +215,6 @@
     right: -50px;
     top: -50px;
     -webkit-transform: rotate(30deg);
-        -ms-transform: rotate(30deg);
```
Comment (AlanFoster) [inline dist/css/react-calendar.css]: Why was this removed?

File: example/css/example.scss [hunk 2]
```diff
@@ -141,3 +141,20 @@ input {
     margin-top: 0;
   }
 }
+.singleDateRange{
```
Comment (AlanFoster) [inline example/css/example.scss]: Mind providing a screenshot? :+1: 

Could we place spaces before the `{` - we don't have a css linter currently! :(

File: dist/css/react-calendar.css (hunk 1 above)
Comment (jonniedarko) [inline dist/css/react-calendar.css]: I would assume this was part of the build, I made no manual changes to the dist folder, just re-ran the build to double check.

File: example/css/example.scss (hunk 2 above)
Comment (jonniedarko) [inline example/css/example.scss]: updated as requested. Here is screen shot as requested:
screen shot

File: example/index.jsx [hunk 3]
```diff
@@ -89,6 +89,55 @@ const DatePickerSingle = React.createClass({
...
+    });
+  },
+
+  render() {
+    return (
+      <div className="singleDateRange">
+        <RangePicker {...this.props} onSelect={this.handleSelect}
+          value={this.state.value} />
+          <div className="buttonContainer">
+            <button type='button' onClick={this.updateDate.bind(this, null)} >today</button>
```
Comment (AlanFoster) [inline example/index.jsx]: Could we change this to `<input type='button'`

File: example/index.jsx [hunk 4]
```diff
@@ -89,6 +89,55 @@ const DatePickerSingle = React.createClass({
...
+            <button type='button' onClick={this.updateDate.bind(this, true, 'month')} >Next Month</button>
```
Comment (AlanFoster) [inline example/index.jsx]: Thoughts on using moment directly?

```js
moment().add(1, 'month')
```

File: example/index.jsx [hunk 5]
```diff
@@ -189,9 +237,16 @@ const Index = React.createClass({
                 selectionType="single"
                 minimumDate={new Date()} />
             </div>
+
+            <div className="example">
+              <h4>Updating calendar date value</h4>
+              <DatePickerSingleWithSetDateButtons
```
Comment (AlanFoster) [inline example/index.jsx]: :+1: for example

File: dist/css/react-calendar.css (hunk 1 above)
Comment (AlanFoster) [inline dist/css/react-calendar.css]: :+1:

File: src/DateRangePicker.jsx [hunk 6]
```diff
@@ -90,11 +90,17 @@ const DateRangePicker = React.createClass({
...
     var nextEnabledRange = this.getEnabledRange(nextProps);
-
-    this.setState({
+    var updatedState = {
       dateStates: this.state.dateStates && Immutable.is(this.state.dateStates, nextDateStates) ? this.state.dateStates : nextDateStates,
       enabledRange: this.state.enabledRange && this.state.enabledRange.isSame(nextEnabledRange) ? this.state.enabledRange : nextEnabledRange,
-    });
+    };
+
+    if (nextProps.value ){
```
Comment (AlanFoster) [inline src/DateRangePicker.jsx]: Going to review this change a bit more; I'm not sure if this works correctly when multiple dates are selected

Comment (jonniedarko) [inline src/DateRangePicker.jsx]: Sure, thanks

Comment (jonniedarko) [inline src/DateRangePicker.jsx]: I actually see what you talking about now, I may need to rework this a little

Comment (jonniedarko) [inline src/DateRangePicker.jsx]: @AlanFoster I am half way towards a solution, as well as being able to set range externally but will not get it complete for a couple of weeks , are you happy to just leave the PR open till then? Or close and I can re-submit when ready if that would be preferable. Thanks

**Comment by AlanFoster** [comment]
@jonniedarko Leaving this PR for now works for me; Let me know when you're good to pick it up again :)

**Comment by AlanFoster** [comment]
@jonniedarko Would you mind adding some tests to this PR too please? :+1:

**Comment by jonniedarko** [comment]
@AlanFoster what would be a prefered way to approach setting the a highlighted range programmatically. Specifically, what do you feel should be the desired result if for example the start and end date are more than a month apart, e.g. if start was feb 1st and end was oct 1st?  should each calendar display the start and end month? or just either the start/end and the month next to it?

**Comment by jonniedarko** [comment]
@AlanFoster Any chance of a review of the changes?

File: dist/calendar/CalendarDate.js [hunk 7]
```diff
@@ -78,23 +78,29 @@ var CalendarDate = _react2['default'].createClass({
...
+  },
 
   mouseUp: function mouseUp() {
     this.props.onSelectDate(this.props.date);
-
-    if (this.state.mouseDown) {
-      this.setState({
-        mouseDown: false
-      });
+    if (!this.isUnmounted) {
```
Comment (AlanFoster) [inline dist/calendar/CalendarDate.js]: We could move this to be a guard clause at the first line

File: dist/calendar/CalendarDate.js [hunk 8]
```diff
@@ -78,23 +78,29 @@ var CalendarDate = _react2['default'].createClass({
       mouseDown: false
     };
   },
+  componentWillUnmount: function componentWillUnmount() {
+    this.isUnmounted = true;
```
Comment (AlanFoster) [inline dist/calendar/CalendarDate.js]: Should we remove our event listeners?

Comment (jonniedarko) [inline dist/calendar/CalendarDate.js]: if we do not, it will try to setState on an unmounted component throwing an error, this prevents that case

File: dist/calendar/CalendarDate.js (hunk 7 above)
Comment (jonniedarko) [inline dist/calendar/CalendarDate.js]: Is it not good practice to cleanup event listeners?

File: dist/calendar/CalendarDate.js (hunk 8 above)
Comment (jonniedarko) [inline dist/calendar/CalendarDate.js]: Is it not good practice to cleanup event listeners?

File: dist/calendar/CalendarDate.js (hunk 7 above)
Comment (jonniedarko) [inline dist/calendar/CalendarDate.js]: :+1:

File: dist/calendar/CalendarDate.js [hunk 9]
```diff
@@ -78,23 +78,29 @@ var CalendarDate = _react2['default'].createClass({
...
     }
-    document.removeEventListener('mouseup', this.mouseUp);
   },
 
   mouseDown: function mouseDown() {
-    this.setState({
-      mouseDown: true
-    });
-    document.addEventListener('mouseup', this.mouseUp);
+    if (!this.isUnmounted) {
```
Comment (AlanFoster) [inline dist/calendar/CalendarDate.js]: Guard clause

Comment (jonniedarko) [inline dist/calendar/CalendarDate.js]: @AlanFoster has been updated - view here does not seem to be showing latest , if you click view full changes you will see it has been updated

**Comment by jonniedarko** [comment]
@AlanFoster any chance of this being merged?

File: dist/calendar/CalendarDate.js (hunk 8 above)
Comment (AlanFoster) [inline dist/calendar/CalendarDate.js]: Out of curiosity, when do we need this `isUnmounted` flag?

File: dist/calendar/CalendarDate.js (hunk 9 above)
Comment (AlanFoster) [inline dist/calendar/CalendarDate.js]: Is this line needed at all? If I understand correctly, we can't trigger a mouseDown event on a component that has been unmounted. Let me know if I'm forgetting an edge case

**Comment by AlanFoster** [comment]
@jonniedarko I've added some code changes as a PR, mind checking them out? https://github.com/jonniedarko/react-daterange-picker/pull/1
If you're happy we could get this merged in

File: dist/calendar/CalendarDate.js (hunk 9 above)
Comment (jonniedarko) [inline dist/calendar/CalendarDate.js]: At the time I  had found that an error was being thrown in the console due to  the event being triggered even though was unmounted, i put this guard in to prevent that error.

**Comment by eliseumds** [comment]
Is the following behaviour expected?



Even though the newly selected range is visible, the calendar moves to the left.

File: dist/calendar/CalendarDate.js (hunk 9 above)
Comment (eliseumds) [inline dist/calendar/CalendarDate.js]: Do we really dynamic events binding/unbinding? The following seems to work (although I didn't test in many browsers):

#### 1. Instead of using state to store *mouseDown*, we can just use an instance property

The state property *mouseDown* does not affect any part of the presentation (UI), so I don't see why it should reside there.

```js
React.createClass({
  isMouseDown: false,

  mouseLeave() {
    if (this.isMouseDown) {
      // ...
    }
  }
});
```

#### 2. Events can be attached directly to the TD element, so we won't have to worry about unmounting

```js
// ...
onTouchEnd={this.touchEnd}
onMouseUp={this.onMouseUp}
// ...
```

**Comment by jonniedarko** [comment]
@eliseumds That behaviour is what I expected but maybe not ideal?

**Comment by eliseumds** [comment]
@jonniedarko ye, it's a bit too much. I really like how [this one](http://www.daterangepicker.com/#examples) behaves:

File: dist/calendar/CalendarDate.js (hunk 9 above)
Comment (jonniedarko) [inline dist/calendar/CalendarDate.js]: @eliseumds While I don't disagree with this observation, I feel this is a separate issue on the approach already in the existing framework, maybe raise a separate issue?

**Comment by jonniedarko** [comment]
@eliseumds I have updated to take you feedback into account regards updating the displayed calendars. @AlanFoster I didn't address his other concern as I felt that should probably be a separate issue, I also updated the example you updated , just to make it a little more obvious to the user what kind of update can be made

**Comment by jonniedarko** [comment]
@AlanFoster FYI I updated the package json peer dependencies to be stricter about choosing react 14, you may wish to update to 15 but you will also need to update the `react-addons-pure-render-mixin` to 15 too, I choose not to update anything other than to make sure they were in sync

**Comment by jonniedarko** [comment]
@AlanFoster having issues dependences after rebasing of the  master, tried updating dependences and tried setting them at react 0.14 but tests failing, can you have a look, i believe the current master branch on your repo is having the same issues

File: dist/DateRangePicker.js [hunk 10]
```diff
@@ -130,11 +130,28 @@ var DateRangePicker = _react2['default'].createClass({
...
+    };
+
+    // if numberOfCalendars are not 2 update as normal
+    if (yearMonth && nextProps.numberOfCalendars !== 2) {
+      updatedState.year = yearMonth.year;
+      updatedState.month = yearMonth.month;
+
+      // prevent updating month & year if there are 2 calendars and the date falls in the 2nd month
+      // this will prevent the calendar moving if within the 2 months already visiable
+    } else if (this.isYearMonthNotCurrentlyDisplayed(yearMonth, nextProps.numberOfCalendars)) {
```
Comment (AlanFoster) [inline dist/DateRangePicker.js]: To remove the duplicate `updatedState` assignment, and the `numberOfCalendars` count assertion

Maybe we could write something like -

```js
if (yearMonth && !this.isYearMonthVisible(yearMonth, nextProps.numberOfCalendars)) {
  updatedState.year = yearMonth.year;
  updatedState.month = yearMonth.month;
} 

this.setState(updatedState);
```

And then maybe update `isYearMonthNotCurrentlyDisplayed` to be re-implemented in a truthy way, something like -

```js
isYearMonthVisible: function isYearMonthVisible(yearMonth, numberOfCalendars) {
  const isSameYear = (yearMonth.year === this.state.year);
  const isMonthVisible =  (yearMonth.month === this.state.month) || (numberOfCalendars === 2 && (yearMonth.month - 1 === this.state.month));
 
  return  isSameYear && isMonthVisible;
}
```

Note: Code hasn't been tested.

**Comment by eliseumds** [comment]
@jonniedarko wow, imo, experience improved a lot! Thank you.

**Comment by AlanFoster** [comment]
@jonniedarko Good catch; I've locked us down to React 0.14
I think we need more work in deciding whether the current year / month is visible. I was also thinking that we might want to jump to the end of the month for a moment range potentially?

**Comment by jonniedarko** [comment]
@AlanFoster What do you mean by jump to end of Month for range? Do you mean show two non consecutive months if Range is more than 2?

**Comment by AlanFoster** [comment]
@jonniedarko i.e. selecting a range which spans multiple months may cause this logic to break unexpectedly, and force it to snap to the very start of your selection again. This would be another variation of @eliseumds's GIF

**Comment by eliseumds** [comment]
@AlanFoster yeah, I guess this is actually how the date picker from Google Analytics behave, but it should only jump like that when there's a new range coming from props, not from user clicks.

**Comment by AlanFoster** [comment]
@jonniedarko @eliseumds I've put up a PR which is an extension of this work, #144 
Would greatly appreciate a PR review if you guys get the chance :+1:

**Comment by AlanFoster** [comment]
@jonniedarko Thanks for the PR; This functionality should now be in master :+1:

**Comment by jonniedarko** [comment]
@AlanFoster nice to see it got in but it kinda sucks not to get the contribution having put the time and work into it. Definitely put me off contributing again!
